learn/building_programs/project_make.md

```diff
@@ -0,0 +1,452 @@
+a robust take on the dependency problem.
+First, we want an approach that can process all source files independently,
+while each source file provides (``module``) or requires (``use``) modules.
+When generating the dependencies only the name of the source file and the
+module files are known, no information on the object file names should be
+required.
+
+If you check the dependency section above you will note that all dependencies are
+defined between object files rather than source files. To change this, we can
+generate a map from the source files their respective object files:
+
+```make
+# Define a map from each file name to its object file
+obj = $(src).o
+$(foreach src, $(SRCS) $(TEST_SRCS), $(eval $(src) := $(obj)))
+```
+
+Note the declaration of ``obj`` as recursively expanded variable, we effectively
+use this mechanism to define a function in ``make``. The ``foreach`` function
+allows us to loop over all source files, while the ``eval`` function allows us
+to generate ``make`` statements and evaluate them for this ``Makefile``.
+
+We adjust the dependencies accordingly as we can now define the name of the
+object files through the source file names:
+
+```make
+# Define all module interdependencies
+csv_kinds.mod := $(src/csv_kinds.f90)
+csv_module.mod := $(src/csv_module.F90)
+csv_parameters.mod := $(src/csv_parameters.f90)
+csv_utilities.mod := $(src/csv_utilities.f90)
+$(src/csv_module.F90): $(csv_utilities.mod)
+$(src/csv_module.F90): $(csv_kinds.mod)
+$(src/csv_module.F90): $(csv_parameters.mod)
+$(src/csv_parameters.f90): $(csv_kinds.mod)
+$(src/csv_utilities.f90): $(csv_kinds.mod)
+$(src/csv_utilities.f90): $(csv_parameters.mod)
+$(src/tests/csv_read_test.f90): $(csv_module.mod)
+$(src/tests/csv_test.f90): $(csv_module.mod)
+$(src/tests/csv_write_test.f90): $(csv_module.mod)
+```
+
+The same strategy of creating a map is already used for the module files, now
+it is just expanded to the object files as well.
+
+To generate the respective dependency map automatically we will use a simple
+``awk`` script here
+
+```awk
+#!/usr/bin/awk -f
+
+BEGIN {
+    IGNORECASE = 1
+}
+
+$1 ~ /^module$/ &&
+$2 ~ /^[a-zA-Z][a-zA-Z1-9_]*$/ {
+    if (modc[FILENAME,$2]++ == 0) {
+        mod[++im] = sprintf("%s.mod = $(%s)", tolower($2), FILENAME)
+    }
+}
+
+$1 ~ /^use$/ &&
+$2 ~ /^[a-zA-Z][a-zA-Z1-9_]*,?$/ {
+    gsub(/,/, "", $2)
+    if (usec[FILENAME,$2]++ == 0) {
+        use[++iu] = sprintf("$(%s) += $(%s.mod)", FILENAME, tolower($2))
+    }
+}
+
+$1 ~ /^(#:?)?include$/ &&
+$2 ~ /^["'].+["']$/ {
+    gsub(/'|"/, "", $2)
+    if (incc[FILENAME,$2]++ == 0) {
+        inc[++ii] = sprintf("$(%s) += %s", FILENAME, $2)
+    }
+}
+
+END {
+    for (i in mod) print mod[i]
+    for (i in use) print use[i]
+    for (i in inc) print inc[i]
+}
+```
+
+This is a particular simple script which makes a few assumptions on the source
+code it is going to parse, so it will certainly not work for all Fortran code,
+but for this example it will suffice.
```

Comment:
```suggestion
This script makes a few assumptions about the source code it parses, so it will not work with all Fortran code, but for this example it will suffice.
```